core/src/trezor/wire/thp/received_message_handler.py

```diff
@@ -0,0 +1,458 @@
+            INIT_HEADER_LENGTH
+            + KEY_LENGTH
+            + TAG_LENGTH : message_length
+            - CHECKSUM_LENGTH
+            - TAG_LENGTH
+        ],
+        0,
+        "ThpHandshakeCompletionReqNoisePayload",
+    )
+    if TYPE_CHECKING:
+        assert ThpHandshakeCompletionReqNoisePayload.is_type_of(noise_payload)
+    enabled_methods = get_enabled_pairing_methods(ctx.iface)
+    for method in noise_payload.pairing_methods:
+        if method not in enabled_methods:
+            raise ThpInvalidDataError()
+        if method not in ctx.selected_pairing_methods:
+            ctx.selected_pairing_methods.append(method)
+    if __debug__ and utils.ALLOW_DEBUG_MESSAGES:
+        log.debug(
+            __name__,
+            "host static pubkey: %s, noise payload: %s",
+            utils.get_bytes_as_str(host_encrypted_static_pubkey),
+            utils.get_bytes_as_str(handshake_completion_request_noise_payload),
+        )
+
+    # key is decoded in handshake._handle_th2_crypto
+    host_static_pubkey = host_encrypted_static_pubkey[:PUBKEY_LENGTH]
+
+    paired: bool = False
+
+    if noise_payload.host_pairing_credential is not None:
+        try:  # TODO change try-except for something better
+            paired = validate_credential(
+                noise_payload.host_pairing_credential,
+                host_static_pubkey,
+            )
+        except DataError as e:
+            if __debug__ and utils.ALLOW_DEBUG_MESSAGES:
+                log.exception(__name__, e)
+            pass
+
+    trezor_state = _TREZOR_STATE_UNPAIRED
+    if paired:
+        trezor_state = _TREZOR_STATE_PAIRED
+    # send hanshake completion response
+    ctx.write_handshake_message(
+        HANDSHAKE_COMP_RES,
+        ctx.handshake.get_handshake_completion_response(trezor_state),
+    )
+
+    ctx.handshake = None
+
+    if paired:
+        ctx.set_channel_state(ChannelState.ENCRYPTED_TRANSPORT)
+    else:
+        ctx.set_channel_state(ChannelState.TP1)
+
+
+async def _handle_state_ENCRYPTED_TRANSPORT(ctx: Channel, message_length: int) -> None:
+    if __debug__ and utils.ALLOW_DEBUG_MESSAGES:
+        log.debug(__name__, "handle_state_ENCRYPTED_TRANSPORT")
+
+    ctx.decrypt_buffer(message_length)
+
+    buffer = memory_manager.get_existing_read_buffer(ctx.get_channel_id_int())
+    # if buffer is BufferError:
+    # pass  # TODO handle
+    session_id, message_type = ustruct.unpack(
+        ">BH", memoryview(buffer)[INIT_HEADER_LENGTH:]
+    )
+    if session_id not in ctx.sessions:
+
+        s = session_manager.get_session_from_cache(ctx, session_id)
+
+        if s is None:
+            s = SeedlessSessionContext(ctx, session_id)
+
+        ctx.sessions[session_id] = s
+        loop.schedule(s.handle())
+
+    elif ctx.sessions[session_id].get_session_state() is SessionState.UNALLOCATED:
+        raise ThpUnallocatedSessionError(session_id)
+        # TODO what should happen here? Should I create a new seedless session?
```

Comment:
Does this happen when client creates a session with passphrase, then it's evicted from cache later? Returning an error seems like the right thing in this case.